Add Variable Expiring Multiparty

PrelaunchTeam opened the pull request.

smb2796 commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,78 @@
+	    
+    }
+
+This function can be called by the UMA governor or the DAO contract if there is a vulnerability or technical problem discovered in the DAO contract. Calling this function successfully will emit the `EmergencyUpdateDAOAddress` event as defined on line 120.
```
In your code you comment on the `variableExpiration` function:

> instead of requesting the oracle price, we set the price as provided by the DAO vote.

Can you explain this more? How is the DAO vote triggered and what DAO are you referring to?

smb2796 replied:
Thinking about this more, I guess the method depends what address is set as the `externalVariableExpirationDAOAddress` right? So this would be flexible and the "DAO" address would just need to pass in the expiry price.

To confirm - this would not require any UMA vote, correct?

Also wondering if you have a DAO contract that you are thinking of using for this?

smb2796 commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,78 @@
+    	    emit VariableExpiration(msg.sender, oldExpirationTimestamp, expirationTimestamp);
+    }
+       
+This function is based on the `emergencyShutdown` function located directly below it. In addition to being accessible by the UMA governor, this function is also accessible by the predefined DAO governance contract. The `settlementPrice` parameter has a maximum of 18 decimals places, and value should be multiplied by 10^18 before being sent to the contract. We also add a new event, `VariableExpira ...
```
Wondering what the rationale behind just passing a price in on VariableExpiration is rather doing something like calling the optimistic oracle?

nicholaspai commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,78 @@
+
+For security purposes in case a vulnerability is discovered with the DAO contract, we include an emergency update function on lines 631 to 636:
+
+    function emergencyUpdateDAOAddress(address DAOAddress) external {
```
You should consider only allowing this function to be set once, presumably upon deployment, therefore there is reassurance to users that the DAO address will not change once set.

pemulis reviewed (commented):
Whoops, this comment should have been submitted last week, but I guess it was left pending. Related to some of the other comments below.

PrelaunchTeam commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,78 @@
+
+  
+
+    function variableExpiration(uint256 settlementPrice) external override onlyPreExpiration() onlyOpenState() nonReentrant() {
```
Yes, we discussed this in the governance meeting on Monday and we will make the update, hopefully this week. Appreciate the feedback.

PrelaunchTeam commented:
https://github.com/PrelaunchFinance/VEMP/blob/main/contracts/PricelessPositionManager.sol#L654
Hey guys so I switched over to the oracle for the price, to deploy it I will need to work on the migration file. 

I looked at this one which has some extra stuff in there for deploying testnet tokens https://github.com/UMAprotocol/protocol/blob/master/packages/core/migrations/15_deploy_expiring_multi_party_creator.js

I will mess around with it more soon, this one is currently returning an out of gas error on ganache https://github.com/PrelaunchFinance/VEMP/blob/main/migrations/2_deploy_variable_expiring_multi_party_creator.js

smb2796 commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,78 @@
+An alternative design where the expiration would be set via UMA vote was considered. We decided to use an application specific governance model to support a wider variety of situations and reduce the burden on UMA voters. 
+
+## Implementation
+The code is available here: https://github.com/PrelaunchFinance/VEMP/blob/main/PricelessPositionManager.sol
```
This link is broken now. Can you update?

smb2796 commented:
> Hey guys so I switched over to the oracle for the price,
Great! Let us know if you need assistance on the migrations file.

Can you update the UMIP to explain the new functionality of the contract?

PrelaunchTeam commented:
> > Hey guys so I switched over to the oracle for the price,
> 
> Great! Let us know if you need assistance on the migrations file.
> 
> Can you update the UMIP to explain the new functionality of the contract?

Ok, updated. Let me know how it looks. Yes if you can help with the migration file for local testing and maybe show me one that was used for deploying to mainnet also that would help

smb2796 commented:
The updates look really good and I don't have any issues with the code except perhaps the contracts should be renamed to differentiate it from the existing ExpiringMultiPartyCreator. Curious what other people like @pemulis and @chrismaree think about the current status?

chrismaree commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+The current implementation of the ExpiringMultiParty (EMP) contract accepts a fixed expiration at the time of creation, which cannot be modified. In our use case, we needed the Prelaunch DAO to be able to expire the contract at a specific unknown future time in response to unpredictable events and factors. With the Variable EMP contract proposed, the Prelaunch DAO will be able to expire the contr ...
+
+## Technical Specification
+The main contract that was modified is `PricelessPositionManager.sol`.
```
if you are going to put code like this within markdown please format it appropriately as a code block. See [markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code-and-syntax-highlighting) reference.

PrelaunchTeam replied:
Done, thanks.

chrismaree commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+    function variableExpiration() external onlyPreExpiration() onlyOpenState() nonReentrant() {
+        require(msg.sender == _getFinancialContractsAdminAddress() || msg.sender == externalVariableExpirationDAOAddress, 'Caller must be the authorized DAO or the UMA governor');
+
+        contractState = ContractState.ExpiredPriceReceived;
```
why is the state set to `ContractState.ExpiredPriceReceived` before the DVM has returned a price?

pemulis replied:
`ContractState.ExpiredPriceRequested` would make more sense.

PrelaunchTeam replied:
Fixed. Good catch!

chrismaree commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+The current implementation of the ExpiringMultiParty (EMP) contract accepts a fixed expiration at the time of creation, which cannot be modified. In our use case, we needed the Prelaunch DAO to be able to expire the contract at a specific unknown future time in response to unpredictable events and factors. With the Variable EMP contract proposed, the Prelaunch DAO will be able to expire the contr ...
+
+## Technical Specification
+The main contract that was modified is `PricelessPositionManager.sol`.
```
consider linking to your modified version. It's this link I believe: https://github.com/PrelaunchFinance/VEMP/blob/main/contracts/PricelessPositionManager.sol

PrelaunchTeam replied:
Done, thanks!

chrismaree commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+
+An alternative design where the expiration would be set via UMA vote was considered. We decided to use an application specific governance model to support a wider variety of situations and reduce the burden on UMA voters. 
+
+## Implementation
```
do we want to consider merging in this into the main UMA protocol repo? it _would_ need to get audited by OZ to accommodate this but that can happen pretty quickly with the continuous audit.

mhairimca replied:
Does this need to happen before the UMIP is passed?

mrice32 replied:
I think the status quo is to _not_ put this in an UMA repo. I think it makes sense to have this code be maintained and owned by the prelaunch team.

PrelaunchTeam replied:
Either way is fine by us. Would the continuous audit still be an option if we maintained this separately in our repo? I doubt we would be able to get an OpenZeppelin audit for this as a new project even if we had lots of money to dispose just because of current demand for audits.

pemulis commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+        require(msg.sender == _getFinancialContractsAdminAddress() || msg.sender == externalVariableExpirationDAOAddress, 'Caller must be the authorized DAO or the UMA governor');
+
+        contractState = ContractState.ExpiredPriceReceived;
+        // Expiratory time now becomes the current time (variable shutdown time).
```
`Expiration time`

PrelaunchTeam replied:
Fixed, though that was taken from the comments on emergencyShutdown in the EMP contract so might want to update that there also eventually

mrice32 replied:
Thanks for pointing this out!!

mhairimca commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+- Author: Prelaunch Team, dev@prelaunch.finance
+- Status: Draft
+- Created: April 14, 2021
+- Discourse Link: TBA (Pending)
```
discourse link?

PrelaunchTeam replied:
Done

PrelaunchTeam commented:
> The UMIP updates look really good and I don't have any issues with the code except perhaps the contracts should be renamed to differentiate it from the existing ExpiringMultiPartyCreator. Curious what other people like @pemulis and @chrismaree think about the current status?

I renamed the contracts

Reinis-FRP commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+The current implementation of the ExpiringMultiParty (EMP) contract accepts a fixed expiration at the time of creation, which cannot be modified. In our use case, we needed the Prelaunch DAO to be able to expire the contract at a specific unknown future time in response to unpredictable events and factors. With the Variable EMP contract proposed, the Prelaunch DAO will be able to expire the contr ...
+
+## Technical Specification
+The main contract that was modified is PricelessPositionManager.sol (Available here: https://github.com/PrelaunchFinance/VEMP/blob/main/contracts/PricelessPositionManager.sol).
```
If you are planning later to merge it to UMA repo, please note that with this [commit](https://github.com/UMAprotocol/protocol/commit/5db0d7178cd6a3c807db4586eeb22a16229e9213) solidity version is upgraded to to 0.8.x and also OZ is upgraded to 4.1.0, so you would need to update your import paths for OZ as they got reorganized.

PrelaunchTeam replied:
I updated the contracts [here](https://github.com/PrelaunchFinance/VEMP/commit/47106e1e03c47f6a962572bb4d2e0055163f0ea5). Is this published yet on NPM? When using @uma/core and trying to compile it said packages were using 0.6.

Reinis-FRP commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+## Technical Specification
+The main contract that was modified is PricelessPositionManager.sol (Available here: https://github.com/PrelaunchFinance/VEMP/blob/main/contracts/PricelessPositionManager.sol).
+
+First, on line 92, we add a new variable:
```
Its  quite difficult to follow changes by line numbers, maybe better just add separate unified diff sections and comment on them?

PrelaunchTeam replied:
Do you know how to do this across two repositories? I tried with git diff on the local files but it's not readable. 
https://github.com/PrelaunchFinance/VEMP/blob/main/diff.md

Reinis-FRP replied:
I thought to include it in the UMIP, but include only the diff snippets already commented here (e.g. no need to show diffs for import statements) just without mentioning line numbers. As an example:

First, we add a new variable:

```
@@ -182,7 +184,8 @@ contract Liquidatable is PricelessPositionManager {
             params.priceFeedIdentifier,
             params.minSponsorTokens,
             params.timerAddress,
-            params.financialProductLibraryAddress
+            params.financialProductLibraryAddress,
+            params.externalVariableExpirationDAOAddress
         )
         nonReentrant()
     {
```

Reinis-FRP replied:
I was just confused with line numbering, so the first comment could be like:

First, we add a new variable:

```
@@ -88,6 +91,9 @@ contract PricelessPositionManager is FeePayer {
     // the functionality of the EMP to support a wider range of financial products.
     FinancialProductLibrary public financialProductLibrary;
 
+    // address for the DAO which will have authority to expire the contract
+    address externalVariableExpirationDAOAddress;
+
     /****************************************
      *                EVENTS                *
      ****************************************/
```

mhairimca commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+## Headers
```
AFAIK, this is waiting for steps to be taken behind the scenes before this can be added to UMA's system.

mhairimca reviewed (commented):
Still waiting for code review and to deploy and verify contracts prior to UMIP vote.

mrice32 commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+
+For security purposes in case a vulnerability is discovered with the DAO contract, we include an emergency update function on lines 631 to 636:
+
+    function emergencyUpdateDAOAddress(address DAOAddress) public {
```
Small nit on naming: I could see this being the result of a normal upgrade. Do you think we should use the term emergency?

PrelaunchTeam replied:
Good point, will update that

mrice32 commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+
+  
+
+    function variableExpiration() external onlyPreExpiration() onlyOpenState() nonReentrant() {
```
Just out of curiosity, what specific use cases are you envisioning? Would it better and more predictable for users if you had a function where the DAO _passed in_ a future timestamp once it's known? Or is that more complicated than it's worth?

PrelaunchTeam replied:
To start the DAO we will use for our governance of our VEMPs will be a multisig council of team, advisors and thought leaders using a gnosis SAFE. I think it's probably a good roadmap item, however one of the other paths we may take towards progressive decentralization would be to make a VEMP which would require an UMA vote to expire, if it is possible with the economics guarantees and CoC incentives. So I think we can work on that if we decide to use the external DAO vote as a long term solution.

mrice32 reviewed (approved):
Looks great!

meirbank commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+        contractState = ContractState.ExpiredPriceRequested;
+        // Expiratory time now becomes the current time (variable shutdown time).
+        // Price received at this time stamp. `settleExpired` can now withdraw at this timestamp.
+        uint256 oldExpirationTimestamp = expirationTimestamp;
```
I guess it's not really needed but shouldn't really have any effect. Tried to change as little as possible.

PrelaunchTeam replied:
There is no specific purpose, just happened as a result of trying to change as little as possible

chrismaree replied:
I guess you could use it to bound how far into the future the variable expiring contract can run for.

chrismaree commented on UMIPs/Add-Variable-Expiring-Multiparty.md:
```diff
@@ -0,0 +1,83 @@
+
+  
+
+    function variableExpiration() external onlyPreExpiration() onlyOpenState() nonReentrant() {
```
it would also be great to correctly format this according to markdown specification see: https://www.markdownguide.org/basic-syntax/#code-blocks-1

meirbank replied:
Updated with 8 spaces. Thanks.

PrelaunchTeam replied:
Updated with 8 spaces. Thanks

smb2796 reviewed (approved):
Thank you for adding the contract addresses! Going to merge in and move to Last Call.
